swarm/pss/snapshot_test.go

```diff
@@ -0,0 +1,433 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	return int(msgCount), int(nodeCount)
+}
+
+func readSnapshot(t *testing.T, nodeCount int) simulations.Snapshot {
+	f, err := os.Open(fmt.Sprintf("testdata/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+	jsonbyte, err := ioutil.ReadAll(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var snap simulations.Snapshot
+	err = json.Unmarshal(jsonbyte, &snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return snap
+}
+
+func assingTestVariables(sim *simulation.Simulation, msgCount int) {
+	log.Debug("-------------------------------------------------------------------------")
+	for _, nodeId := range sim.NodeIDs() {
+		nodeAddrs[nodeId] = nodeIDToAddr(nodeId)
+	}
+
+	for i := 0; i < int(msgCount); i++ {
+		msgAddr := pot.RandomAddress() // we choose message addresses randomly
+		msgs = append(msgs, msgAddr.Bytes())
+		smallestPo := 256
+		var targets []enode.ID
+		var prev int
+
+		// loop through all nodes and add the message to recipient indices
+		for _, nod := range sim.Net.GetNodes() {
+			po, _ := pof(msgs[i], nodeAddrs[nod.ID()], 0)
+			depth := kademlias[nod.ID()].NeighbourhoodDepth()
+
+			// only nodes with closest IDs (wrt msg) will receive the msg
+			if po > prev {
+				prev = po
+				targets = nil
+				targets = append(targets, nod.ID())
+			} else if po == prev {
+				targets = append(targets, nod.ID())
+			}
+
+			if po >= depth {
+				maxMessages++
+				allowed[i] = append(allowed[i], nod.ID())
+				allowedMsgs[nod.ID()] = append(allowedMsgs[nod.ID()], uint64(i))
+			}
+
+			// a node with the smallest PO (wrt msg) will be the sender
+			if po < smallestPo {
+				smallestPo = po
+				senders[i] = nod.ID()
+			}
+		}
+
+		msgsToReceive += len(targets)
+		for _, id := range targets {
+			recipients[i] = append(recipients[i], id)
+			expectedMsgs[id] = append(expectedMsgs[id], uint64(i))
+		}
+
+		log.Debug("nn for msg", "targets", len(recipients[i]), "msgidx", i, "msg", common.Bytes2Hex(msgAddr[:8]), "sender", senders[i], "senderpo", smallestPo)
+	}
+	log.Debug("msgs to receive", "count", msgsToReceive)
+}
+
+func TestProxNetwork(t *testing.T) {
+	if (*runNodes > 0 && *runMessages == 0) || (*runMessages > 0 && *runNodes == 0) {
+		t.Fatal("cannot specify only one of flags --nodes and --messages")
+	} else if *runNodes > 0 {
+		t.Run(fmt.Sprintf("%d/%d", *runMessages, *runNodes), testProxNetwork)
+	} else {
+		t.Run("1/4", testProxNetwork)
+	}
+}
+
+// This tests generates a sequenced number of messages with random addresses.
+// It then calculates which nodes in the network have the address of each message
+// within their nearest neighborhood depth, and stores them as recipients.
+// Upon sending the messages, it verifies that the respective message is passed to the message handlers of these recipients.
+// It will fail if a recipient handles a message it should not, or if after propagation not all expected messages are handled (timeout)
+func testProxNetwork(t *testing.T) {
```

Comment:
your test is functionally LGTM but i feel it could be improved and simplified.

- global vars
- RPC vs API call via bucket object
- snapshot handling?
- flags
- conn labels

Reply:
I concur on global vars, snapshot handling and flags. My opinions given in individual comments.